samples/java-crud-shopping-cart/src/main/java/io/cloudstate/samples/shoppingcart/ShoppingCartEntity.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2019 Lightbend Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudstate.samples.shoppingcart;
+
+import com.example.crud.shoppingcart.Shoppingcart;
+import com.example.crud.shoppingcart.persistence.Domain;
+import com.google.protobuf.Empty;
+import io.cloudstate.javasupport.EntityId;
+import io.cloudstate.javasupport.crud.CommandContext;
+import io.cloudstate.javasupport.crud.CommandHandler;
+import io.cloudstate.javasupport.crud.CrudEntity;
+import io.cloudstate.javasupport.crud.DeleteStateHandler;
+import io.cloudstate.javasupport.crud.UpdateStateHandler;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/** A CRUD entity. */
+@CrudEntity
+public class ShoppingCartEntity {
+  private final String entityId;
+  private final Map<String, Shoppingcart.LineItem> cart = new LinkedHashMap<>();
+
+  public ShoppingCartEntity(@EntityId String entityId) {
+    this.entityId = entityId;
+  }
+
+  @UpdateStateHandler
+  public void handleUpdateState(Domain.Cart cart) {
+    this.cart.clear();
+    for (Domain.LineItem item : cart.getItemsList()) {
+      this.cart.put(item.getProductId(), convert(item));
+    }
+  }
+
+  @DeleteStateHandler
+  public void handleDeleteState() {
+    this.cart.clear();
```

Comment:
@ralphlaude I'm wondering, isn't the annotation sort of redundant since the clear-invocation is used?

Reply:
@viktorklang i think the annotation and the clear-invocation are not redundant. the user function should update its memory state upon deleteEntity which is called in the CommandHandler.

Reply:
@ralphlaude What happens if line 55 is omitted?

Reply:
@viktorklang the user function still has the old state if the line 55 is omitted and is not in sync with the proxy

Reply:
@ralphlaude Having two different deletions (clear() and @DeleteStateHandler) seems like it will easily diverge. Would it not be enough with ctx.updateState(…) and ctx.deleteState()?

Reply:
@viktorklang clear() and @DeleteStateHandler will diverge and it is true. if i understand correctly @UpdateStateHandler and @DeleteStateHandler annotations should be removed. is that correct? instead only ctx.updateState(…) and ctx.deleteState() should be used for managing the state in the user function. to be able to use only context methods the context should have the state passed in and the state can be accessed with ctx.getState() in the user function.
the final result will be CommandContext with methods getState, updateState and deleteState.

Reply:
@ralphlaude Yeah, that was my thinking—then of course you'll need to track multiple invocations of updateState and deleteState in the same logical execution:

```java
//Should this work?
ctx.deleteState();
ctx.updateState(…);
```

and

```java
//Should this work?
ctx.updateState(…);
ctx.deleteState();
```

and

```java
//Should this work?
ctx.deleteState();
ctx.deleteState();
```

and

```java
//Should this work?
ctx.deleteState();
ctx.getState();
```

etc. :)

Reply:
@ralphlaude We discussed this specific point extensively, about subsequent invocations, try to rescue a little bit of that discussion of ours, maybe it can help. Do you remember?

Reply:
@viktorklang :) good to know we are on the same track. i was on another track :). i think all the combinations you mentioned should and will work with the proposed CommandContext. i will change the user facing according to that and the invocations in the combinations should be tracked.

Reply:
@sleipnir i don't remember, may be you can help me on this.

Reply:
This topic was previously discussed in the issue and we talked later on Gitter. 
```
Guy Youansi
@ralphlaude
jun 11 11:33
i will try to explain the problem i am facing in the context of CRUD. let take the shopping cart example an assume the AddLineItem method does a CRUD update and the RemoveLineItem method does a CRUD remove. the shopping cart service has a HTTP endpoints for those method. The CRUD protocol define those create/read/update/remove operations. when the proxy gets a request for this service how would it know which protocol operation to call?
Guy Youansi
@ralphlaude
jun 11 11:44
then i talked to Viktor and he gives some hints. i could be better to have all CRUD operations in the CommandContext
Adriano Santos
@sleipnir
jun 11 11:46
I ended up answering that question in the Issue. But following your example my question would be, how would those methods you mentioned be noted in the entity? Which Note is available to the user. I ask this because I think the user should use annotations to demonstrate what behavior is expected
I saw Viktor's comments in PR, but I definitely don't think this is a good option. Especially because the only thing that differentiates a CRDT from an EventSourced are the annotations used in the entities, to change that would be to break the symmetry of the API's
```

Reply:
https://github.com/cloudstateio/cloudstate/pull/220#discussion_r467947936

Reply:
@sleipnir thanks i remember. the multiple invocations should be handled. the problem was how to handle that according to the user facing API. now all the annotations will be removed that handled the state.

Reply:
@viktorklang :), i was wrong the last time regarding the multiples invocations you mentioned above. i think those multiples invocations should not work at all and a failure should be thrown in such case. Do you agree?
this CRUD stuff is making me crazy :sweat_smile:

Reply:
CRUD: Crazy Read Update Delete 😅

Reply:
@marcellanz :D:D:D:D:D

Reply:
@ralphlaude I don't know! :) I think it is open for debate as to what is the expected semantics for such "progressions"?

Reply:
@viktorklang ok :), it is open for debate and i will bring that up to the contributor call.